### PR TITLE
negative sampling excludes positive class

### PR DESCRIPTION
**Pull request by hoosierEE**
Addressing issue #1228

### CODE REVIEW

File: docs/tutorials/word2vec.ipynb [hunk 1]
```diff
@@ -37,20 +37,20 @@
         "id": "AOpGoE2T-YXS"
```
Comment (cantonios) [inline docs/tutorials/word2vec.ipynb:37]: <p>Can you remove the extra <code>&gt;</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/text/pull/1229/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (cantonios) [inline docs/tutorials/word2vec.ipynb:37]: <p>Same here, there appears to be an extra <code>&gt;</code>.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/text/pull/1229/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='25'/>

Comment (cantonios) [inline docs/tutorials/word2vec.ipynb:37]: <p>"can to sample random words" -&gt; "can sample random words"</p><p> </p><p>Looks like there might be an extra line break between "You can" and "pass words from the positive class"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/text/pull/1229/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='39'/>

Comment (cantonios) [inline docs/tutorials/word2vec.ipynb:37]: <pre class="ql-syntax" spellcheck="false">Line #24.        for i in range(window_size, len(sequence)-window_size):
</pre><p>You may as well build this dict from the existing positive_skip_grams, rather than iterating over the window from scratch.  You could also use a <code>set</code> to avoid duplicate entries. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/text/pull/1229/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='62'/>

Comment (cantonios) [inline docs/tutorials/word2vec.ipynb:37]: <pre class="ql-syntax" spellcheck="false">Line #42.          for target in window[target_word]:
</pre><p>These words are really <code>context</code> rather than <code>target</code>. I think you can simplify that loop as</p><pre class="ql-syntax" spellcheck="false">for sample in negative_sampling_candidates:
  if sample in window[target_word]:
    break
</pre><p>If the window was a set, this lookup would be pretty efficient too.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/text/pull/1229/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='62'/>

Comment (cantonios) [reply docs/tutorials/word2vec.ipynb:37]: Weird, the "ReviewNB" app didn't seem to link it directly to the line....

Line 109.  But looking at the source, it could just be because of the html replacement.  Maybe we should just revert all those changes.

**Comment by cantonios** [comment]
Also, I just reviewed the documentation for [`tf.random.log_uniform_candidate_sampler`](https://www.tensorflow.org/api_docs/python/tf/random/log_uniform_candidate_sampler).  It explicitly states that it does not reject any accidental positive hits, and then links to the [Candidate Sampling Algorithms Reference](https://www.tensorflow.org/extras/candidate_sampling.pdf).  In that reference, the "Negative Sampling" row says that it considers negative training classes to be the full set `S_i`, which does include positive samples.  This is opposed to "Sampled Logistic", which considers the set `(S_i - T_i)`.  So it may be intentional that there could be accidental hits.

File: docs/tutorials/word2vec.ipynb (hunk 1 above)
Comment (hoosierEE) [reply docs/tutorials/word2vec.ipynb:37]: <p>Nice idea, I'll try out something like this:</p><pre class="ql-syntax" spellcheck="false">window = defaultdict(set)
for target_word, context_word in positive_skip_grams:
  window[target_word].add(context_word)
  ...
</pre>

Comment (hoosierEE) [reply docs/tutorials/word2vec.ipynb:37]: <p>Agreed, I'm not sure where that html encoding noise came from. Maybe because I did the edits in colab? I'll revert all the ones I find.</p>

Comment (hoosierEE) [reply docs/tutorials/word2vec.ipynb:37]: <p>Thanks for catching these, will fix.</p>

**Comment by hoosierEE** [comment]
Your feedback was very helpful, thanks! Building a set from the `positive_skip_grams` ended up being much faster. On average I see around 10% of negative samples discarded because they overlap with the positive context, resulting in about 1 percentage point improvement in training accuracy at 20 epochs.
